File: dotnet/src/Vulgarity/Model/VulgarityMatch.cs
using System;

namespace Vulgarity
{
    /// <summary>One term found in the text.</summary>
    /// <remarks>
    /// <see cref="Start"/> and <see cref="End"/> index the ORIGINAL text the
    /// caller passed in, not the folded form. A caller can highlight the real
    /// span with them.
    /// </remarks>
    public sealed class VulgarityMatch : IComparable<VulgarityMatch>
    {
        internal VulgarityMatch(int start, int end, VulgarityTerm term, int termIndex)
        {
            Start = start;
            End = end;
            Term = term;
            TermIndex = termIndex;
        }

        /// <summary>Where the match starts in the original text.</summary>
        public int Start { get; private set; }

        /// <summary>Where the match ends in the original text. Exclusive.</summary>
        public int End { get; private set; }

        /// <summary>How many characters of the original text the match covers.</summary>
        public int Length
        {
            get { return End - Start; }
        }

        /// <summary>The term that matched.</summary>
        public VulgarityTerm Term { get; private set; }

        /// <summary>The position of the term in the filter's term list.</summary>
        internal int TermIndex { get; private set; }

        /// <summary>The matched term, folded to profile fold-v1.</summary>
        public string Text
        {
            get { return Term.Text; }
        }

        public VulgarityCategory Category
        {
            get { return Term.Category; }
        }

        public string CategoryName
        {
            get { return Term.CategoryName; }
        }

        public int Severity
        {
            get { return Term.Severity; }
        }

        /// <summary>Returns the exact text the match covers.</summary>
        public string Excerpt(string original)
        {
            return original.Substring(Start, End - Start);
        }

        /// <summary>Orders matches by start, then by the longer match, then by term.</summary>
        public int CompareTo(VulgarityMatch other)
        {
            if (other == null)
            {
                return 1;
            }

            if (Start != other.Start)
            {
                return Start < other.Start ? -1 : 1;
            }

            if (End != other.End)
            {
                return End > other.End ? -1 : 1;
            }

            return TermIndex.CompareTo(other.TermIndex);
        }

        public override string ToString()
        {
            return Term.Text + "@" + Start + ".." + End;
        }
    }
}

File: dart/lib/src/model/vulgarity_match.dart
import 'vulgarity_category.dart';
import 'vulgarity_term.dart';

/// One term found in the text.
///
/// [start] and [end] index the ORIGINAL text the caller passed in, not the
/// folded form. A caller can highlight the real span with them.
class VulgarityMatch implements Comparable<VulgarityMatch> {
  const VulgarityMatch(this.start, this.end, this.term, this.termIndex);

  /// Where the match starts in the original text.
  final int start;

  /// Where the match ends in the original text. Exclusive.
  final int end;

  /// The term that matched.
  final VulgarityTerm term;

  /// The position of the term in the filter's term list.
  final int termIndex;

  /// How many characters of the original text the match covers.
  int get length => end - start;

  /// The matched term, folded to profile fold-v1.
  String get text => term.text;

  VulgarityCategory get category => term.category;

  String get categoryName => term.categoryName;

  int get severity => term.severity;

  /// Returns the exact text the match covers.
  String excerpt(String original) => original.substring(start, end);

  /// Orders matches by start, then by the longer match, then by term.
  @override
  int compareTo(VulgarityMatch other) {
    if (start != other.start) {
      return start < other.start ? -1 : 1;
    }
    if (end != other.end) {
      return end > other.end ? -1 : 1;
    }
    return termIndex.compareTo(other.termIndex);
  }

  @override
  String toString() => '${term.text}@$start..$end';
}
